### DISCUSSION TITLE
Data replication in Kademlia

### DISCUSSION THREAD

**Discussion by mtbitcr** [libp2p/rust-libp2p#5677] [Q&A] [answered]
How are things going with adding data replication in Kademlia DHT? Is there anything planned for the near future?

**Comment by guillaumemichel** [comment]
You can read about Kademlia replication in the [specs](https://github.com/libp2p/specs/blob/master/kad-dht/README.md#replication-parameter-k).

Currently, by default all data written to the DHT is replicated on `20` peers. Are you referring to something else?

**Reply by mtbitcr** [reply]
Thank you for your reply. 
I have something else in mind. Probably the best analogy would be torrent. I would like to store files on multiple nodes and if one node goes offline, automatically copy the file to the new node.

**Reply by guillaumemichel** [reply, marked as answer]
If you use the `PUT_VALUE` operation (see [specs](https://github.com/libp2p/specs/blob/master/kad-dht/README.md#dht-operations)), the value you write will be replicated to `20` peers by default. If you want the value stored to persist over time, the node that advertised it in the first place needs to [republish](https://github.com/libp2p/specs/blob/master/kad-dht/README.md#content-provider-advertisement-and-discovery) it periodically, so that even if nodes go offline, the value is always stored on the closest peers and can be discovered by others.

If you want a system where nodes leaving the network would allocate their values to another node, or the DHT servers reallocating values by themselves, you would have to implement it yourself

**Reply by mtbitcr** [reply]
Thanks, @guillaumemichel !
